internal/monitors/statsd/monitor.go

```diff
@@ -0,0 +1,137 @@
+package statsd
+
+import (
+	"context"
+	"time"
+
+	"github.com/signalfx/golib/datapoint"
+	"github.com/signalfx/golib/sfxclient"
+	"github.com/signalfx/signalfx-agent/internal/core/config"
+	"github.com/signalfx/signalfx-agent/internal/monitors"
+	"github.com/signalfx/signalfx-agent/internal/monitors/types"
+	"github.com/signalfx/signalfx-agent/internal/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+const monitorType = "statsd"
+
+var logger = utils.NewThrottledLogger(log.WithFields(log.Fields{"monitorType": monitorType}), 30*time.Second)
+var golibLogger = &utils.LogrusGolibShim{FieldLogger: logger.FieldLogger}
+
+func init() {
+	monitors.Register(monitorType, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+type converterInput struct {
+	Pattern    string
+	Metric     string
+	Dimensions []string
+}
+
+// Config for this monitor
+type Config struct {
+	config.MonitorConfig `yaml:",inline" acceptsEndpoints:"false" singleInstance:"true"`
+	// The host/address on which to bind the UDP listener that accepts statsd
+	// datagrams
+	ListenAddress string `yaml:"listenAddress" default:"localhost"`
+	// The port on which to listen for statsd messages
+	ListenPort uint16 `yaml:"listenPort" default:"8125"`
+	// The protocol for statsd communications
+	Protocol string `yaml:"protocol" default:"udp"`
+	// HTTP timeout duration for both read and writes. This should be a
+	// duration string that is accepted by https://golang.org/pkg/time/#ParseDuration
+	ServerTimeout  time.Duration    `yaml:"serverTimeout" default:"5s"`
+	DeleteSets     bool             `yaml:"deleteSets"`
+	DeleteCounters bool             `yaml:"deleteCounters"`
+	DeleteTimers   bool             `yaml:"deleteTimers"`
+	DeleteGauges   bool             `yaml:"deleteGauges"`
+	MetricPrefix   string           `yaml:"metricPrefix"`
+	Converters     []converterInput `yaml:"converters"`
+}
+
+// Monitor that accepts and forwards trace spans
+type Monitor struct {
+	Output types.Output
+	cancel context.CancelFunc
+	conf   *Config
+}
+
+// Configure the monitor and kick off volume metric syncing
+func (m *Monitor) Configure(conf *Config) error {
+	var ctx context.Context
+	ctx, m.cancel = context.WithCancel(context.Background())
+
+	m.conf = conf
+
+	var converters []*converter
+	for _, ci := range conf.Converters {
+		converters = append(converters, initConverter(&converterInput{
+			Pattern:    ci.Pattern,
+			Metric:     ci.Metric,
+			Dimensions: ci.Dimensions,
+		}))
+	}
+
+	listener := &statsDListener{
+		ipAddr:     conf.ListenAddress,
+		port:       conf.ListenPort,
+		protocol:   conf.Protocol,
+		prefix:     conf.MetricPrefix,
+		converters: converters,
+	}
+
+	err := listener.Listen()
+	if err != nil {
+		return err
+	}
+
+	utils.RunOnInterval(ctx, func() {
+		metrics, err := listener.FetchMetrics()
```

Comment:
The listener should be continuously reacting to new UDP datagrams (and TCP streams) such that it pulls data off of the socket immediately, instead of letting it buffer for many seconds and then flushing the whole thing at once.  Otherwise, you are relying on the kernel to buffer things and you have little to no control over how much it will do before dropping things.